Document MSRV

Is there a minimal supported Rust version for this crate? It seems to require associated constants, which makes it incompatible with older Rust versions.

This is the error:

```
error[E0599]: no associated item named `MIN` found for type `i8` in the current scope
   --> /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/easy-cast-0.2.0/src/lib.rs:319:46
    |
319 |                     const LBOUND: $x = <$y>::MIN as $x;
    |                                              ^^^ associated item not found in `i8`
...
353 | impl_float!(f32: i8, i16, i32, i64, i128, isize);
    | ------------------------------------------------- in this macro invocation
```

So I think it is about using the associated constant in `std`, which is not available in old Rust versions. Should be easy to fix (by importing the module), but I don't know whether you want to bother to support such old Rust versions.

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -10,13 +10,17 @@ jobs:
   test:
     name: Test
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        toolchain: ["beta", "1.36.0"]
     steps:
       - uses: actions/checkout@v2
       - name: Install toolchain
         uses: actions-rs/toolchain@v1
         with:
           profile: minimal
-          toolchain: nightly
+          toolchain: ${{ matrix.toolchain }}
           override: true
           components: rustfmt
       - name: Rustfmt check

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,12 @@
 Changelog
 =========
 
+## [0.3.0] — 2021-03-29
+
+-   Explicitly support Rust 1.36.0 (and potentially older; #10)
+-   Support `no_std` (#10)
+-   Fix rounding for `floor` on negative values (#10)
+
 ## [0.2.0] — 2021-03-20
 
 -   Add feature flags controlling assert behaviour

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "easy-cast"
-version = "0.2.0"
+version = "0.3.0"
 authors = ["Diggory Hardy <git@dhardy.name>"]
 edition = "2018"
 license = "Apache-2.0"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -11,6 +11,11 @@ keywords = ["cast", "into", "from", "conversion"]
 repository = "https://github.com/kas-gui/easy-cast"
 
 [features]
+default = ["std"]
+
+# Without std, float conversions are disabled
+std = []
+
 # Note: assertions are always used in debug builds; these only affect release builds:
 
 # Always use all assertions

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,7 +13,7 @@ repository = "https://github.com/kas-gui/easy-cast"
 [features]
 default = ["std"]
 
-# Without std, float conversions are disabled
+# Without std, float conversions are disabled (unless libm is used)
 std = []
 
 # Note: assertions are always used in debug builds; these only affect release builds:
@@ -29,3 +29,8 @@ assert_non_neg = []
 
 # Always assert value is within target type's range
 assert_range = []
+
+[dependencies.libm]
+# libm may be used instead of std to provide float conversions
+version = "0.2.1"
+optional = true

--- a/README.md
+++ b/README.md
@@ -30,6 +30,11 @@ If the `always_assert` feature flag is set, assertions will be turned on in
 all builds. Some additional feature flags are available for finer-grained
 control (see [Cargo.toml](Cargo.toml)).
 
+## MSRV
+
+The Minumum Supported Rust Version is 1.36.0 (older versions may work but are
+untested).
+
 
 Copyright and Licence
 -------

--- a/README.md
+++ b/README.md
@@ -37,6 +37,7 @@ untested).
 
 By default, `std` support is required. With default features disabled `no_std`
 is supported, but the `ConvFloat` and `CastFloat` traits are unavailable.
+Enabling the `libm` feature will re-enable these traits.
 
 
 Copyright and Licence

--- a/README.md
+++ b/README.md
@@ -30,10 +30,13 @@ If the `always_assert` feature flag is set, assertions will be turned on in
 all builds. Some additional feature flags are available for finer-grained
 control (see [Cargo.toml](Cargo.toml)).
 
-## MSRV
+## MSRV and no_std
 
 The Minumum Supported Rust Version is 1.36.0 (older versions may work but are
 untested).
+
+By default, `std` support is required. With default features disabled `no_std`
+is supported, but the `ConvFloat` and `CastFloat` traits are unavailable.
 
 
 Copyright and Licence

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -29,6 +29,8 @@
 //! all builds. Some additional feature flags are available for finer-grained
 //! control (see `Cargo.toml`).
 
+#![deny(missing_docs)]
+
 use std::mem::size_of;
 
 /// Like [`From`], but supporting potentially-fallible conversions
@@ -51,6 +53,7 @@ use std::mem::size_of;
 ///
 /// [`From`]: std::convert::From
 pub trait Conv<T> {
+    /// Convert from `T` to `Self` (see trait doc)
     fn conv(v: T) -> Self;
 }
 
@@ -381,6 +384,7 @@ impl ConvFloat<f32> for u128 {
 
 /// Like [`Into`], but for [`Conv`]
 pub trait Cast<T> {
+    /// Cast from `Self` to `T` (see trait doc)
     fn cast(self) -> T;
 }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -31,6 +31,7 @@
 
 #![deny(missing_docs)]
 #![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(doc_cfg, feature(doc_cfg))]
 
 use core::mem::size_of;
 
@@ -298,6 +299,7 @@ impl_via_digits_check!(usize: f32, f64);
 /// If the source value is out-of-range or not-a-number then the conversion must
 /// fail with a panic.
 #[cfg(feature = "std")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
 pub trait ConvFloat<T> {
     /// Convert to the nearest integer
     ///
@@ -404,6 +406,7 @@ impl<S, T: Conv<S>> Cast<T> for S {
 
 /// Like [`Into`], but for [`ConvFloat`]
 #[cfg(feature = "std")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
 pub trait CastFloat<T> {
     /// Cast to the nearest integer
     ///

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -332,6 +332,7 @@ macro_rules! impl_float {
             }
             #[inline]
             fn conv_floor(x: $x) -> $y {
+                let x = x.floor();
                 if cfg!(any(debug_assertions, feature = "assert_float")) {
                     const LBOUND: $x = std::$y::MIN as $x;
                     const UBOUND: $x = std::$y::MAX as $x + 1.0;

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -111,17 +111,17 @@ impl_via_as_neg_check!(i128: u128);
 
 // Assumption: $y::MAX is representable as $x
 macro_rules! impl_via_as_max_check {
-    ($x:ty: $y:ty) => {
+    ($x:ty: $y:tt) => {
         impl Conv<$x> for $y {
             #[inline]
             fn conv(x: $x) -> $y {
                 #[cfg(any(debug_assertions, feature = "assert_range"))]
-                assert!(x <= <$y>::MAX as $x);
+                assert!(x <= std::$y::MAX as $x);
                 x as $y
             }
         }
     };
-    ($x:ty: $y:ty, $($yy:ty),+) => {
+    ($x:ty: $y:tt, $($yy:tt),+) => {
         impl_via_as_max_check!($x: $y);
         impl_via_as_max_check!($x: $($yy),+);
     };
@@ -136,17 +136,17 @@ impl_via_as_max_check!(u128: u8, u16, u32, u64);
 
 // Assumption: $y::MAX and $y::MIN are representable as $x
 macro_rules! impl_via_as_range_check {
-    ($x:ty: $y:ty) => {
+    ($x:ty: $y:tt) => {
         impl Conv<$x> for $y {
             #[inline]
             fn conv(x: $x) -> $y {
                 #[cfg(any(debug_assertions, feature = "assert_range"))]
-                assert!(<$y>::MIN as $x <= x && x <= <$y>::MAX as $x);
+                assert!(std::$y::MIN as $x <= x && x <= std::$y::MAX as $x);
                 x as $y
             }
         }
     };
-    ($x:ty: $y:ty, $($yy:ty),+) => {
+    ($x:ty: $y:tt, $($yy:tt),+) => {
         impl_via_as_range_check!($x: $y);
         impl_via_as_range_check!($x: $($yy),+);
     };
@@ -158,22 +158,22 @@ impl_via_as_range_check!(i64: i8, i16, i32, u8, u16, u32);
 impl_via_as_range_check!(i128: i8, i16, i32, i64, u8, u16, u32, u64);
 
 macro_rules! impl_int_signed_dest {
-    ($x:ty: $y:ty) => {
+    ($x:ty: $y:tt) => {
         impl Conv<$x> for $y {
             #[inline]
             fn conv(x: $x) -> $y {
                 if size_of::<$x>() == size_of::<$y>() {
                     #[cfg(any(debug_assertions, feature = "assert_range"))]
-                    assert!(x <= <$y>::MAX as $x);
+                    assert!(x <= std::$y::MAX as $x);
                 } else if size_of::<$x>() > size_of::<$y>() {
                     #[cfg(any(debug_assertions, feature = "assert_range"))]
-                    assert!(<$y>::MIN as $x <= x && x <= <$y>::MAX as $x);
+                    assert!(std::$y::MIN as $x <= x && x <= std::$y::MAX as $x);
                 }
                 x as $y
             }
         }
     };
-    ($x:ty: $y:ty, $($yy:ty),+) => {
+    ($x:ty: $y:tt, $($yy:tt),+) => {
         impl_int_signed_dest!($x: $y);
         impl_int_signed_dest!($x: $($yy),+);
     };
@@ -193,7 +193,7 @@ impl_int_signed_dest!(isize: i8, i16, i32, i64, i128);
 impl_int_signed_dest!(usize: i8, i16, i32, i64, i128, isize);
 
 macro_rules! impl_int_signed_to_unsigned {
-    ($x:ty: $y:ty) => {
+    ($x:ty: $y:tt) => {
         impl Conv<$x> for $y {
             #[inline]
             fn conv(x: $x) -> $y {
@@ -201,13 +201,13 @@ macro_rules! impl_int_signed_to_unsigned {
                 assert!(x >= 0);
                 if size_of::<$x>() > size_of::<$y>() {
                     #[cfg(any(debug_assertions, feature = "assert_range"))]
-                    assert!(x <= <$y>::MAX as $x);
+                    assert!(x <= std::$y::MAX as $x);
                 }
                 x as $y
             }
         }
     };
-    ($x:ty: $y:ty, $($yy:ty),+) => {
+    ($x:ty: $y:tt, $($yy:tt),+) => {
         impl_int_signed_to_unsigned!($x: $y);
         impl_int_signed_to_unsigned!($x: $($yy),+);
     };
@@ -221,19 +221,19 @@ impl_int_signed_to_unsigned!(i128: usize);
 impl_int_signed_to_unsigned!(isize: u8, u16, u32, u64, u128, usize);
 
 macro_rules! impl_int_unsigned_to_unsigned {
-    ($x:ty: $y:ty) => {
+    ($x:ty: $y:tt) => {
         impl Conv<$x> for $y {
             #[inline]
             fn conv(x: $x) -> $y {
                 if size_of::<$x>() > size_of::<$y>() {
                     #[cfg(any(debug_assertions, feature = "assert_range"))]
-                    assert!(x <= <$y>::MAX as $x);
+                    assert!(x <= std::$y::MAX as $x);
                 }
                 x as $y
             }
         }
     };
-    ($x:ty: $y:ty, $($yy:ty),+) => {
+    ($x:ty: $y:tt, $($yy:tt),+) => {
         impl_int_unsigned_to_unsigned!($x: $y);
         impl_int_unsigned_to_unsigned!($x: $($yy),+);
     };
@@ -257,21 +257,21 @@ impl Conv<f64> for f32 {
 }
 
 macro_rules! impl_via_digits_check {
-    ($x:ty: $y:ty) => {
+    ($x:ty: $y:tt) => {
         impl Conv<$x> for $y {
             #[inline]
             fn conv(x: $x) -> $y {
                 if cfg!(any(debug_assertions, feature = "assert_digits")) {
                     let src_ty_bits = u32::conv(size_of::<$x>() * 8);
                     let src_digits = src_ty_bits - (x.leading_zeros() + x.trailing_zeros());
-                    let dst_digits = <$y>::MANTISSA_DIGITS;
+                    let dst_digits = std::$y::MANTISSA_DIGITS;
                     assert!(src_digits <= dst_digits);
                 }
                 x as $y
             }
         }
     };
-    ($x:ty: $y:ty, $($yy:ty),+) => {
+    ($x:ty: $y:tt, $($yy:tt),+) => {
         impl_via_digits_check!($x: $y);
         impl_via_digits_check!($x: $($yy),+);
     };
@@ -309,15 +309,15 @@ pub trait ConvFloat<T> {
 }
 
 macro_rules! impl_float {
-    ($x:ty: $y:ty) => {
+    ($x:ty: $y:tt) => {
         impl ConvFloat<$x> for $y {
             #[inline]
             fn conv_nearest(x: $x) -> $y {
                 let x = x.round();
                 if cfg!(any(debug_assertions, feature = "assert_float")) {
                     // Tested: these limits work for $x=f32 and all $y except u128
-                    const LBOUND: $x = <$y>::MIN as $x;
-                    const UBOUND: $x = <$y>::MAX as $x + 1.0;
+                    const LBOUND: $x = std::$y::MIN as $x;
+                    const UBOUND: $x = std::$y::MAX as $x + 1.0;
                     assert!(x >= LBOUND && x < UBOUND);
                 }
                 x as $y
@@ -325,8 +325,8 @@ macro_rules! impl_float {
             #[inline]
             fn conv_floor(x: $x) -> $y {
                 if cfg!(any(debug_assertions, feature = "assert_float")) {
-                    const LBOUND: $x = <$y>::MIN as $x;
-                    const UBOUND: $x = <$y>::MAX as $x + 1.0;
+                    const LBOUND: $x = std::$y::MIN as $x;
+                    const UBOUND: $x = std::$y::MAX as $x + 1.0;
                     assert!(x >= LBOUND && x < UBOUND);
                 }
                 x as $y
@@ -335,15 +335,15 @@ macro_rules! impl_float {
             fn conv_ceil(x: $x) -> $y {
                 let x = x.ceil();
                 if cfg!(any(debug_assertions, feature = "assert_float")) {
-                    const LBOUND: $x = <$y>::MIN as $x;
-                    const UBOUND: $x = <$y>::MAX as $x + 1.0;
+                    const LBOUND: $x = std::$y::MIN as $x;
+                    const UBOUND: $x = std::$y::MAX as $x + 1.0;
                     assert!(x >= LBOUND && x < UBOUND);
                 }
                 x as $y
             }
         }
     };
-    ($x:ty: $y:ty, $($yy:ty),+) => {
+    ($x:ty: $y:tt, $($yy:tt),+) => {
         impl_float!($x: $y);
         impl_float!($x: $($yy),+);
     };

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -30,8 +30,9 @@
 //! control (see `Cargo.toml`).
 
 #![deny(missing_docs)]
+#![cfg_attr(not(feature = "std"), no_std)]
 
-use std::mem::size_of;
+use core::mem::size_of;
 
 /// Like [`From`], but supporting potentially-fallible conversions
 ///
@@ -51,7 +52,7 @@ use std::mem::size_of;
 /// Note that you *may not* want to use this where loss of precision is
 /// acceptable, e.g. if an approximate conversion `x as f64` suffices.
 ///
-/// [`From`]: std::convert::From
+/// [`From`]: core::convert::From
 pub trait Conv<T> {
     /// Convert from `T` to `Self` (see trait doc)
     fn conv(v: T) -> Self;
@@ -119,7 +120,7 @@ macro_rules! impl_via_as_max_check {
             #[inline]
             fn conv(x: $x) -> $y {
                 #[cfg(any(debug_assertions, feature = "assert_range"))]
-                assert!(x <= std::$y::MAX as $x);
+                assert!(x <= core::$y::MAX as $x);
                 x as $y
             }
         }
@@ -144,7 +145,7 @@ macro_rules! impl_via_as_range_check {
             #[inline]
             fn conv(x: $x) -> $y {
                 #[cfg(any(debug_assertions, feature = "assert_range"))]
-                assert!(std::$y::MIN as $x <= x && x <= std::$y::MAX as $x);
+                assert!(core::$y::MIN as $x <= x && x <= core::$y::MAX as $x);
                 x as $y
             }
         }
@@ -167,10 +168,10 @@ macro_rules! impl_int_signed_dest {
             fn conv(x: $x) -> $y {
                 if size_of::<$x>() == size_of::<$y>() {
                     #[cfg(any(debug_assertions, feature = "assert_range"))]
-                    assert!(x <= std::$y::MAX as $x);
+                    assert!(x <= core::$y::MAX as $x);
                 } else if size_of::<$x>() > size_of::<$y>() {
                     #[cfg(any(debug_assertions, feature = "assert_range"))]
-                    assert!(std::$y::MIN as $x <= x && x <= std::$y::MAX as $x);
+                    assert!(core::$y::MIN as $x <= x && x <= core::$y::MAX as $x);
                 }
                 x as $y
             }
@@ -204,7 +205,7 @@ macro_rules! impl_int_signed_to_unsigned {
                 assert!(x >= 0);
                 if size_of::<$x>() > size_of::<$y>() {
                     #[cfg(any(debug_assertions, feature = "assert_range"))]
-                    assert!(x <= std::$y::MAX as $x);
+                    assert!(x <= core::$y::MAX as $x);
                 }
                 x as $y
             }
@@ -230,7 +231,7 @@ macro_rules! impl_int_unsigned_to_unsigned {
             fn conv(x: $x) -> $y {
                 if size_of::<$x>() > size_of::<$y>() {
                     #[cfg(any(debug_assertions, feature = "assert_range"))]
-                    assert!(x <= std::$y::MAX as $x);
+                    assert!(x <= core::$y::MAX as $x);
                 }
                 x as $y
             }
@@ -267,7 +268,7 @@ macro_rules! impl_via_digits_check {
                 if cfg!(any(debug_assertions, feature = "assert_digits")) {
                     let src_ty_bits = u32::conv(size_of::<$x>() * 8);
                     let src_digits = src_ty_bits - (x.leading_zeros() + x.trailing_zeros());
-                    let dst_digits = std::$y::MANTISSA_DIGITS;
+                    let dst_digits = core::$y::MANTISSA_DIGITS;
                     assert!(src_digits <= dst_digits);
                 }
                 x as $y
@@ -296,6 +297,7 @@ impl_via_digits_check!(usize: f32, f64);
 ///
 /// If the source value is out-of-range or not-a-number then the conversion must
 /// fail with a panic.
+#[cfg(feature = "std")]
 pub trait ConvFloat<T> {
     /// Convert to the nearest integer
     ///
@@ -311,6 +313,7 @@ pub trait ConvFloat<T> {
     fn conv_ceil(x: T) -> Self;
 }
 
+#[cfg(feature = "std")]
 macro_rules! impl_float {
     ($x:ty: $y:tt) => {
         impl ConvFloat<$x> for $y {
@@ -353,11 +356,16 @@ macro_rules! impl_float {
 }
 
 // Assumption: usize < 128-bit
+#[cfg(feature = "std")]
 impl_float!(f32: i8, i16, i32, i64, i128, isize);
+#[cfg(feature = "std")]
 impl_float!(f32: u8, u16, u32, u64, usize);
+#[cfg(feature = "std")]
 impl_float!(f64: i8, i16, i32, i64, i128, isize);
+#[cfg(feature = "std")]
 impl_float!(f64: u8, u16, u32, u64, u128, usize);
 
+#[cfg(feature = "std")]
 impl ConvFloat<f32> for u128 {
     #[inline]
     fn conv_nearest(x: f32) -> u128 {
@@ -395,6 +403,7 @@ impl<S, T: Conv<S>> Cast<T> for S {
 }
 
 /// Like [`Into`], but for [`ConvFloat`]
+#[cfg(feature = "std")]
 pub trait CastFloat<T> {
     /// Cast to the nearest integer
     ///
@@ -410,6 +419,7 @@ pub trait CastFloat<T> {
     fn cast_ceil(self) -> T;
 }
 
+#[cfg(feature = "std")]
 impl<S, T: ConvFloat<S>> CastFloat<T> for S {
     fn cast_nearest(self) -> T {
         T::conv_nearest(self)

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -291,6 +291,37 @@ impl_via_digits_check!(u128: f32, f64);
 impl_via_digits_check!(isize: f32, f64);
 impl_via_digits_check!(usize: f32, f64);
 
+#[cfg(all(not(feature = "std"), feature = "libm"))]
+trait FloatRound {
+    fn round(self) -> Self;
+    fn floor(self) -> Self;
+    fn ceil(self) -> Self;
+}
+#[cfg(all(not(feature = "std"), feature = "libm"))]
+impl FloatRound for f32 {
+    fn round(self) -> Self {
+        libm::roundf(self)
+    }
+    fn floor(self) -> Self {
+        libm::floorf(self)
+    }
+    fn ceil(self) -> Self {
+        libm::ceilf(self)
+    }
+}
+#[cfg(all(not(feature = "std"), feature = "libm"))]
+impl FloatRound for f64 {
+    fn round(self) -> Self {
+        libm::round(self)
+    }
+    fn floor(self) -> Self {
+        libm::floor(self)
+    }
+    fn ceil(self) -> Self {
+        libm::ceil(self)
+    }
+}
+
 /// Nearest / floor / ceil conversions from floating point types
 ///
 /// This trait is explicitly for conversions from floating-point values to
@@ -298,8 +329,8 @@ impl_via_digits_check!(usize: f32, f64);
 ///
 /// If the source value is out-of-range or not-a-number then the conversion must
 /// fail with a panic.
-#[cfg(feature = "std")]
-#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+#[cfg(any(feature = "std", feature = "libm"))]
+#[cfg_attr(doc_cfg, doc(cfg(any(feature = "std", feature = "libm"))))]
 pub trait ConvFloat<T> {
     /// Convert to the nearest integer
     ///
@@ -315,7 +346,8 @@ pub trait ConvFloat<T> {
     fn conv_ceil(x: T) -> Self;
 }
 
-#[cfg(feature = "std")]
+#[cfg(any(feature = "std", feature = "libm"))]
+#[cfg_attr(doc_cfg, doc(cfg(any(feature = "std", feature = "libm"))))]
 macro_rules! impl_float {
     ($x:ty: $y:tt) => {
         impl ConvFloat<$x> for $y {
@@ -324,8 +356,8 @@ macro_rules! impl_float {
                 let x = x.round();
                 if cfg!(any(debug_assertions, feature = "assert_float")) {
                     // Tested: these limits work for $x=f32 and all $y except u128
-                    const LBOUND: $x = std::$y::MIN as $x;
-                    const UBOUND: $x = std::$y::MAX as $x + 1.0;
+                    const LBOUND: $x = core::$y::MIN as $x;
+                    const UBOUND: $x = core::$y::MAX as $x + 1.0;
                     assert!(x >= LBOUND && x < UBOUND);
                 }
                 x as $y
@@ -334,8 +366,8 @@ macro_rules! impl_float {
             fn conv_floor(x: $x) -> $y {
                 let x = x.floor();
                 if cfg!(any(debug_assertions, feature = "assert_float")) {
-                    const LBOUND: $x = std::$y::MIN as $x;
-                    const UBOUND: $x = std::$y::MAX as $x + 1.0;
+                    const LBOUND: $x = core::$y::MIN as $x;
+                    const UBOUND: $x = core::$y::MAX as $x + 1.0;
                     assert!(x >= LBOUND && x < UBOUND);
                 }
                 x as $y
@@ -344,8 +376,8 @@ macro_rules! impl_float {
             fn conv_ceil(x: $x) -> $y {
                 let x = x.ceil();
                 if cfg!(any(debug_assertions, feature = "assert_float")) {
-                    const LBOUND: $x = std::$y::MIN as $x;
-                    const UBOUND: $x = std::$y::MAX as $x + 1.0;
+                    const LBOUND: $x = core::$y::MIN as $x;
+                    const UBOUND: $x = core::$y::MAX as $x + 1.0;
                     assert!(x >= LBOUND && x < UBOUND);
                 }
                 x as $y
@@ -359,16 +391,16 @@ macro_rules! impl_float {
 }
 
 // Assumption: usize < 128-bit
-#[cfg(feature = "std")]
+#[cfg(any(feature = "std", feature = "libm"))]
 impl_float!(f32: i8, i16, i32, i64, i128, isize);
-#[cfg(feature = "std")]
+#[cfg(any(feature = "std", feature = "libm"))]
 impl_float!(f32: u8, u16, u32, u64, usize);
-#[cfg(feature = "std")]
+#[cfg(any(feature = "std", feature = "libm"))]
 impl_float!(f64: i8, i16, i32, i64, i128, isize);
-#[cfg(feature = "std")]
+#[cfg(any(feature = "std", feature = "libm"))]
 impl_float!(f64: u8, u16, u32, u64, u128, usize);
 
-#[cfg(feature = "std")]
+#[cfg(any(feature = "std", feature = "libm"))]
 impl ConvFloat<f32> for u128 {
     #[inline]
     fn conv_nearest(x: f32) -> u128 {
@@ -406,8 +438,8 @@ impl<S, T: Conv<S>> Cast<T> for S {
 }
 
 /// Like [`Into`], but for [`ConvFloat`]
-#[cfg(feature = "std")]
-#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+#[cfg(any(feature = "std", feature = "libm"))]
+#[cfg_attr(doc_cfg, doc(cfg(any(feature = "std", feature = "libm"))))]
 pub trait CastFloat<T> {
     /// Cast to the nearest integer
     ///
@@ -423,7 +455,7 @@ pub trait CastFloat<T> {
     fn cast_ceil(self) -> T;
 }
 
-#[cfg(feature = "std")]
+#[cfg(any(feature = "std", feature = "libm"))]
 impl<S, T: ConvFloat<S>> CastFloat<T> for S {
     fn cast_nearest(self) -> T {
         T::conv_nearest(self)

--- a/tests/tests.rs
+++ b/tests/tests.rs
@@ -21,5 +21,5 @@ fn float_casts() {
 #[test]
 #[should_panic]
 fn u32_max_f32() {
-    f32::conv(u32::MAX);
+    f32::conv(std::u32::MAX);
 }

--- a/tests/tests.rs
+++ b/tests/tests.rs
@@ -17,6 +17,7 @@ fn float_casts() {
     assert_eq!(x, 14);
     assert_eq!(u8::conv_floor(13.8f64), 13);
     assert_eq!(u32::conv_ceil(13.1f32), 14);
+    assert_eq!(i64::conv_floor(-3168565.13), -3168566);
 }
 
 #[test]

--- a/tests/tests.rs
+++ b/tests/tests.rs
@@ -10,6 +10,7 @@ fn int_casts() {
 }
 
 #[test]
+#[cfg(feature = "std")]
 fn float_casts() {
     assert_eq!(u64::conv_nearest(13.2f32), 13);
     let x: i128 = 13.5f32.cast_nearest();
@@ -21,5 +22,5 @@ fn float_casts() {
 #[test]
 #[should_panic]
 fn u32_max_f32() {
-    f32::conv(std::u32::MAX);
+    f32::conv(core::u32::MAX);
 }

